File: 200-NumberOfIslands/200-NumberOfIslands.py
# Last updated: 21.02.2026, 21:00:10
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        
        visited = set()
        n = len(grid)
        m = len(grid[0])
        moves = [(-1,0),(1,0),(0,1),(0,-1)]
        islands = 0
        for x in range(n):
            for y in range(m):
                if grid[x][y] == "0" or (x,y) in visited:
                    continue
                visited.add((x,y)) 
                queue = [(x,y)]                
                ind = 0
                islands += 1
                while ind<len(queue):
                    curr_x, curr_y = queue[ind]
                    ind += 1
                    visited.add((curr_x,curr_y))
                    for dx,dy in moves:
                        new_x,new_y = curr_x+dx,curr_y+dy
                        if (0 <= new_x < n) and (0 <= new_y < m):
                            if (new_x, new_y) not in visited and grid[new_x][new_y] == '1':
                                visited.add((new_x, new_y))
                                queue.append((new_x, new_y))

        return islands

File: 200-NumberofIslands/200-NumberofIslands.cpp
// Last updated: 21.02.2026, 21:07:53
1#include <tuple>
2#include <vector>
3class Solution {
4public:
5    vector<tuple<int, int>> moves {make_tuple(-1,0),make_tuple(0,-1),make_tuple(1,0),make_tuple(0,1)};
6    void dfs(vector<vector<char>>& grid ,int x, int y,int n,int m){
7        if (x < 0  || x >= n || y < 0 || y >= m || grid[x][y] != '1'){
8            return;
9        }
10        grid[x][y] = 'v';
11        
12        for (tuple<int,int> move : moves){
13            int dx= x+get<0>(move);
14            int dy = y+get<1>(move);
15            dfs(grid,dx,dy,n,m);
16        }
17    }
18
19    int numIslands(vector<vector<char>>& grid) {
20        int m = grid[0].size();
21        int n = grid.size();
22        int island = 0;
23        for (int i = 0; i<n ; i++){
24            for (int j =0; j<m ; j++){
25                if (grid[i][j] == '1'){
26                    island++;
27                    dfs(grid,i,j,n,m);
28                }
29            }
30        }   
31        return island   ;
32    }
33};